File: Homework/Hw4/matrixHw4.java

public class matrixHw4{

    public static int[][] multiply(int[][] A, int[][] B) {
        if (A[0].length != B.length) {
            throw new IllegalArgumentException("Matrix multiplication is not defined.");
        }

        int rowsA = A.length;
        int colsA = A[0].length;
        int colsB = B[0].length;
        int[][] result = new int[rowsA][colsB];
        
        for (int i = 0; i < rowsA; i++) {
            for (int j = 0; j < colsB; j++) {
                for (int k = 0; k < colsA; k++) {
                    result[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        
        return result;
    }

    // Time Complexity: O(n^3) for square matrices of size n.
}

File: Homework/Hw4/matrixHw4.py

def matrixHw4(A, B):
    if len(A[0]) != len(B):
        raise ValueError("Matrix multiplication is not defined.")

    rows_A = len(A)
    cols_A = len(A[0])
    cols_B = len(B[0])
    result = [[0] * cols_B for _ in range(rows_A)]
    
    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                result[i][j] += A[i][k] * B[k][j]
    return result

# Time Complexity: O(n^3) for square matrices of size n.
